File: server/src/main/kotlin/eu/mlich/lucy/model/money/TransactionCategory.kt
package eu.mlich.lucy.model.money

import java.util.*
import javax.persistence.*

/**
 * Represent a category of transactions.
 *
 * @author Juraj Mlich <jurajmlich@gmail.com>
 */
@Entity
@Table(name = "transaction_category")
data class TransactionCategory(
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        @Column(name = "id")
        var id: Int? = null,

        @Column(name = "name")
        var name: String,

        @Column(name = "color")
        var color: String? = null,

        /**
         * Used primarily for expenses?
         */
        @Column(name = "negative")
        var negative: Boolean,

        @Column(name = "disabled")
        var disabled: Boolean,

        @Column(name = "public_key")
        var publicKey: UUID = UUID.randomUUID()

) {
    override fun equals(other: Any?) = other is Deposit && other.publicKey === publicKey

    override fun hashCode(): Int = publicKey.hashCode()
}

File: android/lib/model/transaction_category.dart
import 'package:android/model/id_aware.dart';

class TransactionCategory extends IdAware<String> {
  String id;
  String name;
  String color;
  bool negative;
  bool disabled;

  TransactionCategory(this.id);

  @override
  String toString() {
    return 'Transaction{id: $id, name: $name, color: $color, negative: $negative, disabled: $disabled}';
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TransactionCategory &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;
}
